pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package versioned
+
+import (
+	"math"
+	"slices"
+	"sync/atomic"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type Version uint64
+
+const (
+	// invalidVersion is never found from the value range.
+	// Also used as the upper bound for new values, so that the value is
+	// found when looking with 'maxVersion'
+	invalidVersion = Version(math.MaxUint64)
+
+	// maxVersion in a handle finds all non-removed values
+	maxVersion = Version(math.MaxUint64 - 1)
+
+	maxEmptyHandles = 1024
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "versioned")
+)
+
+// Handle is the object needed for getting values containing the latest version available at the
+// time of getting the Handle.
+type Handle struct {
+	// Handle stores version + 1 so that the zero value translates to invalidVersion
+	versionPlus1 Version
+}
+
+func newHandle(version Version) Handle {
+	return Handle{versionPlus1: Version(version) + 1}
+}
+
+// AllHandle returns a handle that matches all current/non-removed values
+func AllHandle() Handle {
+	return newHandle(maxVersion)
+}
+
+func (h Handle) getVersion() Version {
+	return h.versionPlus1 - 1
+}
+
+func (h Handle) IsValid() bool {
+	return h.versionPlus1 != 0
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) Load() Version {
+	return Version(a.version.Load())
+}
+
+func (a *atomicVersion) Store(version Version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) CompareAndSwap(old, new Version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+// Manager manages the versioning for multiple VersionedValues.
+// It is to be stored wherever VersionedValues are stored
```

Comment:
What does this mean?

Reply:
Should read `versioned.Value`s. But generally `Manager` is used to share a single version number space across multiple `versioned.Value`s.